src/compiler/commandLineParser.ts

```diff
@@ -584,4 +581,768 @@ namespace ts {
+
+        // Wildcard directories (provided as part of a wildcard path) are stored in a
+        // file map that marks whether it was a regular wildcard match (with a `*` or `?` token),
+        // or a recursive directory. This information is used by filesystem watchers to monitor for
+        // new entries in these paths.
+        const wildcardDirectories = createFileMap<WatchDirectoryFlags>(keyMapper);
+
+        // To reduce the overhead of disk I/O (and marshalling to managed code when hosted in
+        // Visual Studio), file system queries are cached during the expansion session.
+        // If present, a cache entry can be one of three values:
+        // - A `false` value indicates the file or directory did not exist.
+        // - A `true` value indicates the path is a file and it exists.
+        // - An object value indicates the path is a directory and exists. The object may have
+        //   zero, one, or both of the following properties:
+        //   - A "files" array, which contains the file names in the directory.
+        //   - A "directories" array, which contains the subdirectory names in the directory.
+        const cache = createFileMap<boolean | DirectoryResult>(keyMapper);
+
+        // Rather than requery this for each file and filespec, we querythe supported extensions
+        // once and store it on the expansion context.
+        const supportedExtensions = getSupportedExtensions(options);
+
+        // The expansion context holds references to shared information for the various expansion
+        // operations to reduce the overhead of closures.
+        const context: ExpansionContext = {
+            options,
+            host,
+            errors,
+            excludePattern,
+            literalFiles,
+            wildcardFiles,
+            wildcardDirectories,
+            supportedExtensions,
+            cache
+        };
+
+        // Literal files are always included verbatim. An "include" or "exclude" specification cannot
+        // remove a literal file.
+        if (fileNames) {
+            for (const fileName of fileNames) {
+                const path = toPath(fileName, basePath, caseSensitiveKeyMapper);
+                if (!literalFiles.contains(path)) {
+                    literalFiles.set(path, path);
+                }
+            }
+        }
+
+        // Each "include" specification is expanded and matching files are added.
+        if (includeSpecs) {
+            for (let includeSpec of includeSpecs) {
+                includeSpec = normalizePath(includeSpec);
+                includeSpec = removeTrailingDirectorySeparator(includeSpec);
+                expandFileSpec(includeSpec, <Path>basePath, 0, context);
+            }
+        }
+
+        return {
+            fileNames: wildcardFiles.reduce(addFileToOutput, literalFiles.reduce(addFileToOutput, [])),
+            wildcardDirectories: wildcardDirectories.reduce<Map<WatchDirectoryFlags>>(addDirectoryToOutput, {}),

```

Comment:
Isn't `reduce` going pass the index of the current item to `addDirectoryToOutput` for its `directory` parameter?